server/api/users.js

```diff
@@ -274,26 +276,69 @@ internals.applyRoutes = function (server, next) {
         },
         handler: function (request, reply) {
 
-            const id = request.params.id;
-            const update = {
-                $set: {
-                    isActive: request.payload.isActive,
-                    username: request.payload.username,
-                    email: request.payload.email
-                }
-            };
+            Async.auto({
+                user: function (done) {
 
-            User.findByIdAndUpdate(id, update, (err, user) => {
+                    const id = request.params.id;
+                    const update = {
+                        $set: {
+                            isActive: request.payload.isActive,
+                            username: request.payload.username,
+                            email: request.payload.email
+                        }
+                    };
+
+                    User.findByIdAndUpdate(id, update, done);
+                },
+                admin: ['user', function (results, done) {
+
+                    if (!results.user || !results.user.roles ) {
+                        return done();
+                    }
+                    const { admin } = results.user.roles;
+                    if (!admin) {
+                        return done();
+                    }
+                    const update = {
+                        $set: {
+                            user: {
+                                id: results.user._id,
+                                name: request.payload.username
+                            }
+                        }
+                    };
+                    Admin.findByIdAndUpdate(admin.id, update, done);
```

Comment:
This `admin` task (and the `account` task below) could be reduced to something like:

```js
const filter = {
	'user.id': request.params.id
};

const update = {
    $set: {
        user: {
            id: request.params.id,
            name: request.payload.username
        }
    }
};

Admin.findOneAndUpdate(filter, update, done);
```

The results of these Admin and Account updates don't really matter. If a record exits it will be updated and if one doesn't nothing happens.

We've also reduced the surface area for testing by removing a lot of if/else branches.

This also removes the need to lookup the User. We should be able to just just fire off all the update commands and do no lookups.

Reply:
I added a new commit using this approach it simplified the code a lot.
I wasn't aware of the `findOneAndUpdate` function should have checked the `mongo-models` api.